exchanges/stream/websocket.go

```diff
@@ -1292,3 +1103,149 @@ func drain(ch <-chan error) {
+func (w *Websocket) observeData(dropped *int) (exit bool) {
+	select {
+	case <-w.ShutdownC:
+		return true
+	case d := <-w.DataHandler:
+		select {
+		case w.ToRoutine <- d:
+			if *dropped != 0 {
+				log.Infof(log.WebsocketMgr, "%s exchange websocket ToRoutine channel buffer recovered; %d messages were dropped", w.exchangeName, dropped)
+				*dropped = 0
+			}
+		default:
+			if *dropped == 0 {
+				// If this becomes prone to flapping we could drain the buffer, but that's extreme and we'd like to avoid it if possible
+				log.Warnf(log.WebsocketMgr, "%s exchange websocket ToRoutine channel buffer full; dropping messages", w.exchangeName)
+			}
+			*dropped++
+		}
+		return false
+	}
+}
+
+// monitorConnection monitors the connection and attempts to reconnect if the connection is lost
+func (w *Websocket) monitorConnection() func() bool {
+	timer := time.NewTimer(w.connectionMonitorDelay)
+	return func() bool { return w.observeConnection(timer) }
+}
+
+// observeConnection observes the connection and attempts to reconnect if the connection is lost
+func (w *Websocket) observeConnection(t *time.Timer) (exit bool) {
+	select {
+	case err := <-w.ReadMessageErrors:
+		if errors.Is(err, errConnectionFault) {
+			log.Warnf(log.WebsocketMgr, "%v websocket has been disconnected. Reason: %v", w.exchangeName, err)
+			if w.IsConnected() {
+				if shutdownErr := w.Shutdown(); shutdownErr != nil {
+					log.Errorf(log.WebsocketMgr, "%v websocket: connectionMonitor shutdown err: %s", w.exchangeName, shutdownErr)
+				}
+			}
+		}
+		// Speedier reconnection, instead of waiting for the next cycle.
+		if w.IsEnabled() && (!w.IsConnected() && !w.IsConnecting()) {
+			if connectErr := w.Connect(); connectErr != nil {
+				log.Errorln(log.WebsocketMgr, connectErr)
+			}
+		}
+		w.DataHandler <- err // hand over the error to the data handler (shutdown and reconnection is priority)
+	case <-t.C:
+		if w.verbose {
+			log.Debugf(log.WebsocketMgr, "%v websocket: running connection monitor cycle", w.exchangeName)
+		}
+		if !w.IsEnabled() {
+			if w.verbose {
+				log.Debugf(log.WebsocketMgr, "%v websocket: connectionMonitor - websocket disabled, shutting down", w.exchangeName)
+			}
+			if w.IsConnected() {
+				if err := w.Shutdown(); err != nil {
+					log.Errorln(log.WebsocketMgr, err)
+				}
+			}
+			if w.verbose {
+				log.Debugf(log.WebsocketMgr, "%v websocket: connection monitor exiting", w.exchangeName)
+			}
+			t.Stop()
+			w.connectionMonitorRunning.Store(false)
+			return true
+		}
+		if !w.IsConnecting() && !w.IsConnected() {
+			err := w.Connect()
+			if err != nil {
+				log.Errorln(log.WebsocketMgr, err)
+			}
+		}
+		t.Reset(w.connectionMonitorDelay)
+	}
+	return false
+}
+
+// monitorTraffic monitors to see if there has been traffic within the trafficTimeout time window. If there is no traffic
+// the connection is shutdown and will be reconnected by the connectionMonitor routine.
+func (w *Websocket) monitorTraffic() func() bool {
```

Comment:
I tried simulating a disconnect on Binance and got the following:
```
panic: close of closed channel

goroutine 950 [running]:
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*WebsocketConnection).Shutdown(0x14000b12000)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket_connection.go:275 +0x70
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).shutdown(0x14000264800)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:553 +0x364
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).Shutdown(0x0?)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:512 +0xb8
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).observeTraffic.func1()
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:1233 +0x24
created by github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).observeTraffic in goroutine 427
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:1232 +0x1b4
exit status 2
```

This does not occur on `master` branch. I have not been able to re-do this

Reply:
I had a different recreation on closing the application:
```
^C[INFO]  | SYNC | 30/09/2024 12:09:56 | GateIO websocket BTC-USDT SPOT ORDERBOOK: Bids len: 1 Amount: 2.933270 BTC. Total value: 189864.407233 Asks len: 1 Amount: 0.045910 BTC. Total value: 2971.66248
[INFO]  | LOG | 30/09/2024 12:09:56 | Captured interrupt, shutdown requested.
[DEBUG] | LOG | 30/09/2024 12:09:56 | Engine shutting down..
[DEBUG] | ORDER | 30/09/2024 12:09:56 | Order manager shutting down...
[DEBUG] | PORTFOLIO | 30/09/2024 12:09:56 | Portfolio manager shutting down...
[DEBUG] | CONNECTION | 30/09/2024 12:09:56 | Connection manager shutting down...
[DEBUG] | CONNECTION | 30/09/2024 12:09:56 | Connection manager stopped.
[DEBUG] | PORTFOLIO | 30/09/2024 12:09:56 | Portfolio manager shutdown.
[DEBUG] | ORDER | 30/09/2024 12:09:56 | Order manager shutdown.
[DEBUG] | DISPATCH | 30/09/2024 12:09:56 | Dispatch manager shutting down...
[DEBUG] | DISPATCH | 30/09/2024 12:09:56 | Dispatch manager shutdown
[DEBUG] | EXCHANGE | 30/09/2024 12:09:56 | Currency state manager shutting down...
[DEBUG] | EXCHANGE | 30/09/2024 12:09:56 | Currency state manager shutdown.
panic: close of closed channel

goroutine 677 [running]:
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*WebsocketConnection).Shutdown(0x14000c06000)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket_connection.go:275 +0x70
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).shutdown(0x140001fc800)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:553 +0x364
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).Shutdown(0x14000406be0?)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:512 +0xb8
github.com/thrasher-corp/gocryptotrader/exchanges.(*Base).Shutdown(0x1400035a008)
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/exchange.go:1644 +0x2c
github.com/thrasher-corp/gocryptotrader/engine.(*ExchangeManager).Shutdown.func1(0x14000cfe040, 0x14000cfe030, {0x1017f2e20, 0x1400035a008})
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/engine/exchange_manager.go:159 +0x3c
created by github.com/thrasher-corp/gocryptotrader/engine.(*ExchangeManager).Shutdown in goroutine 1
	/Users/WIF/go/src/github.com/thrasher-corp/gocryptotrader/engine/exchange_manager.go:158 +0x1f8
exit status 2
```